src/Main.elm

```diff
@@ -0,0 +1,139 @@
+module Main exposing (..)
+
+import Html exposing (..)
+import Html.App as App
+import Html.Attributes exposing (..)
+import Html.Events exposing (..)
+import Json.Decode as Json
+import String
+
+
+-- main
+
+main : Program Never
+main =
+  App.beginnerProgram
+    { model = emptyModel
+    , view = view
+    , update = update
+    }
+
+-- Model
+
+type alias Model =
+  { entries : List Entry
+  , input : String
+  , uid : Int
+  }
+
+type alias Entry =
+  { description : String
+  , id : Int
+  , taps : Int
+  }
+
+emptyModel : Model
+emptyModel =
+  { entries = []
+  , input = ""
+  , uid = 0
+  }
+
+
+-- Update
+
+type Msg
+  = NoOp
+  | UpdateInput String
+  | Add
+  | Tap Int
+
+
+update : Msg -> Model -> Model
+update msg model =
+  case msg of
+    NoOp -> model
+
+    UpdateInput str ->
+      { model | input = str }
+
+    Add ->
+      { model
+        | input = ""
+        , uid = model.uid + 1
+        , entries =
+            if String.isEmpty model.input then
+              model.entries
+            else
+              model.entries ++ [newEntry model.input model.uid]}
+
+    Tap id' ->
+      let updateEntry e =
+        if e.id == id' then { e | taps = e.taps + 1 } else e
+      in
+        { model | entries = List.map updateEntry model.entries }
+
+
+newEntry : String -> Int -> Entry
+newEntry desc id' =
+  { description = desc
+  , id = id'
+  , taps = 0
+  }
+
+onEnter : Msg -> Attribute Msg
+onEnter msg =
+  let
+    tagger code =
+      if code == 13 then msg else NoOp
+  in
+    on "keydown" (Json.map tagger keyCode)
+
+
+-- View
+
+view : Model -> Html Msg
+view model =
+  div
+    []
+    [ header
+        []
+        [ h1 [] [ text "PRIORITIES" ]
+        , div
+          [ class "input-container" ]
+          [ input
+              [ placeholder "What's your idea?"
+              , autofocus True
+              , value model.input
+              , onInput UpdateInput
+              , onEnter Add
+              ]
+              []
+          ]
+        ]
+    , main'
+        []
+        [ div
+            [ id "todos" ] <|
+            List.map viewEntry model.entries
+        , div
+            [ class "instructions" ]
+            [ p [] [ text "Tap to prioritize." ]
+            -- , p [] [ text "Tap-hold to remove priority." ]
+            ]
+        ]
+    ]
+
+viewEntry : Entry -> Html Msg
+viewEntry entry =
+  div
+    [ class "todo"
+    , onClick <| Tap entry.id
```

Comment:
It seems like all these "event" functions have a type signature like `Msg -> Attribute Msg`; so `Tap entry.id` has type `Attribute Msg`. (`Tap entry.id` will already be evaluated when `viewEntry` is called (I think?), so it doesn't get called when a click event happens.)

So in other words, these event functions take a value of type `Msg` and convert it into something that can be part of the list of attributes in one of Elm's `Html` functions, which have types like:

    div : List (Attribute a) -> List (Html a) -> Html a

I don't really understand the `Html` module yet, so I'm not sure what's going on here. I'm guessing that a value of type `Html a` is something like … a representation of the stream of values of type `a` that's emitted by a DOM subtree. So `main` takes `view : Model -> Html Msg`, applies it to the infinite stream of `Model` values that represents the changing app state, and then pipes the resulting stream of `Msg` values to the function `update : Msg -> Model -> Model`, resulting in more changes to the model.